### PR TITLE
Relabel with GPT-4

### PR DESCRIPTION
**Pull request by arcticfly**
### Changes
* Add dataset entry provenance
* Add dataset entry history
* Add job to relabel with GPT-4
* Show relabeling status in dataset entries table

### CODE REVIEW

File: app/prisma/schema.txt [hunk 1]
```diff
@@ -0,0 +1,697 @@
+// This is your Prisma schema file,
```
Comment (corbt) [inline app/prisma/schema.txt]: rm

File: app/prisma/migrations/20231031174505_add_relabel_request/migration.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+-- CreateEnum
+CREATE TYPE "RelabelRequestStatus" AS ENUM ('PENDING', 'IN_PROGRESS', 'COMPLETE', 'ERROR');
+
+-- Add new columns with default values
+ALTER TABLE "DatasetEntry"
+ADD COLUMN "importId" TEXT NOT NULL DEFAULT '',
+ADD COLUMN "provenance" "DatasetEntryProvenance" NOT NULL DEFAULT 'UPLOAD';
+
+-- Update existing rows to use the default values
+UPDATE "DatasetEntry" SET "importId" = '' WHERE "importId" IS NULL;
```
Comment (corbt) [inline app/prisma/migrations/20231031174505_add_relabel_request/migration.sql:20]: Nit but I think this line and the next are actually unnecessary, the `DEFAULT` that you set when creating the column should already take care of this. No need to change it now though.

File: app/src/components/datasets/DatasetContentTabs/General/DatasetEntriesTable/DatasetEntryEditorDrawer.tsx [hunk 3]
```diff
@@ -106,67 +111,120 @@ export default function DatasetEntryEditorDrawer({
...
+                  <>
+                    <Text>
+                      <Text as="span" fontWeight="bold">
+                        Import ID:
+                      </Text>{" "}
+                      {datasetEntry.importId}
+                    </Text>
+                  </>
+                )}
+                <Text>
```
Comment (corbt) [inline app/src/components/datasets/DatasetContentTabs/General/DatasetEntriesTable/DatasetEntryEditorDrawer.tsx:148]: We might be able to both fit the visual styling of the drawer better and take up less space by removing the outline/different-colored background but not blocking, I can play with that later.

File: app/src/components/datasets/DatasetContentTabs/General/DatasetEntriesTable/DatasetEntriesTable.tsx [hunk 4]
```diff
@@ -1,33 +1,55 @@
...
 import { TableHeader, TableRow, EmptyTableRow } from "./TableRow";
 import DatasetEntryEditorDrawer from "./DatasetEntryEditorDrawer";
 
 export default function DatasetEntriesTable() {
   const [expandedDatasetEntryId, setExpandedDatasetEntryId] = useState<string | null>(null);
-  const datasetEntries = useDatasetEntries().data?.entries;
+  const [refetchInterval, setRefetchInterval] = useState(0);
+  const datasetEntries = useDatasetEntries(refetchInterval).data?.entries;
+
+  const relabelingVisible = useMemo(() => {
```
Comment (corbt) [inline app/src/components/datasets/DatasetContentTabs/General/DatasetEntriesTable/DatasetEntriesTable.tsx:14]: Nice touch, the column helps a lot with making it clear what's going on.

File: app/src/server/api/routers/datasetEntries.router.ts [hunk 5]
```diff
@@ -466,6 +498,46 @@ export const datasetEntriesRouter = createTRPCRouter({
...
+        },
+      });
+
+      if (!dataset) {
+        return error("Dataset not found for dataset entry");
+      }
+
+      await requireCanModifyProject(dataset.projectId, ctx);
+
+      const batchId = Date.now().toString();
```
Comment (corbt) [inline app/src/server/api/routers/datasetEntries.router.ts]: Can we use `new Date().toISOString()` here instead? Will be more interpretable when it shows up in the filters.

File: app/src/server/api/routers/datasetEntries.router.ts [hunk 6]
```diff
@@ -71,6 +72,13 @@ export const datasetEntriesRouter = createTRPCRouter({
             "de.updatedAt as updatedAt",
             "de.outdated as outdated",
             "de.datasetId as datasetId",
+            jsonArrayFrom(
```
Comment (corbt) [inline app/src/server/api/routers/datasetEntries.router.ts:75]: This seems good enough; probably will only have O(1) relabel requests per entry anyway so not worth optimizing.

File: app/src/server/api/routers/datasetEntries.router.ts [hunk 7]
```diff
@@ -299,7 +322,14 @@ export const datasetEntriesRouter = createTRPCRouter({
         })
         .filter(truthyFilter);
 
-      const datasetEntriesToCreate = await prepareDatasetEntriesForImport(datasetId, rowsToConvert);
+      const importId = Date.now().toString();
```
Comment (corbt) [inline app/src/server/api/routers/datasetEntries.router.ts]: `new Date().toISOString()` for readability

File: app/src/server/tasks/importDatasetEntries.task.ts [hunk 8]
```diff
@@ -79,11 +80,15 @@ export const importDatasetEntries = defineTask<ImportDatasetEntriesJob>({
       },
     });
 
+    const importId = Date.now().toString();
```
Comment (corbt) [inline app/src/server/tasks/importDatasetEntries.task.ts]: `new Date().toISOString()` for readability.

File: app/src/server/utils/prepareDatasetEntriesForImport.ts [hunk 9]
```diff
@@ -29,7 +32,10 @@ export const prepareDatasetEntriesForImport = async (
   existingTrainingCount += entriesToImport.filter((row) => row.split === "TRAIN").length;
 
   const newTotalEntries = existingCount + entriesToImport.length;
-  const numTrainingToAdd = Math.floor(trainingRatio * newTotalEntries) - existingTrainingCount;
+  const numTrainingToAdd = Math.min(
```
Comment (corbt) [inline app/src/server/utils/prepareDatasetEntriesForImport.ts]: Curious what drove this change -- I don't immediately see the bug in the code as it was written previously but maybe I'm missing something?

**Review by corbt — CHANGES_REQUESTED** [review]
Minor changes but nice work, excited to have this available!

File: app/prisma/migrations/20231031174505_add_relabel_request/migration.sql (hunk 2 above)
Comment (arcticfly) [reply app/prisma/migrations/20231031174505_add_relabel_request/migration.sql:20]: Yes you're right. I forgot to remove these lines after editing an earlier version.

File: app/src/server/utils/prepareDatasetEntriesForImport.ts (hunk 9 above)
Comment (arcticfly) [reply app/src/server/utils/prepareDatasetEntriesForImport.ts]: It fixed a bug in the previous version of the code, but I didn't see how much you had changed. I think the changes you made fixed the bug that this is supposed to fix, so I can remove it now.

File: app/src/server/utils/prepareDatasetEntriesForImport.ts [hunk 10]
```diff
@@ -26,12 +29,10 @@ export const prepareDatasetEntriesForImport = async (
 
   const trainingRatio = dataset?.trainingRatio ?? 0.8;
 
-  existingTrainingCount += entriesToImport.filter((row) => row.split === "TRAIN").length;
-
   const newTotalEntries = existingCount + entriesToImport.length;
   const numTrainingToAdd = Math.floor(trainingRatio * newTotalEntries) - existingTrainingCount;
 
-  let numTrainingAdded = 0;
+  let numTrainingAdded = entriesToImport.filter((row) => row.split === "TRAIN").length;
```
Comment (arcticfly) [inline app/src/server/utils/prepareDatasetEntriesForImport.ts:35]: I think it's more clear to set `numTrainingAdded` to actually encompass the total number of training rows added in this call. Same goes for `existingTrainingCount`, it should be equal to the number of training rows that existed before this route was called.

File: app/src/server/api/routers/datasetEntries.router.ts (hunk 5 above)
Comment (arcticfly) [reply app/src/server/api/routers/datasetEntries.router.ts]: Works for me

File: app/src/server/utils/prepareDatasetEntriesForImport.ts (hunk 10 above)
Comment (corbt) [reply app/src/server/utils/prepareDatasetEntriesForImport.ts:35]: I agree that with your new changes the code is cleaner than the way I wrote it initially.

**Review by corbt — APPROVED** [review]
Nice work.

**Pull request merged**